torch/csrc/api/src/nn/modules/conv.cpp

```diff
@@ -20,119 +25,163 @@ ConvImpl<D, Derived>::ConvImpl(const ConvOptions<D>& options_)
+    std::vector<int64_t> weight_sizes = {
+      *options.out_channels(),
+      *options.in_channels() / options.groups()};
+    weight_sizes.insert(weight_sizes.end(), (*options.kernel_size().value()).begin(), (*options.kernel_size().value()).end());
+    weight = this->register_parameter(
+      "weight",
+      torch::empty(weight_sizes));
   }
-  weights_size.insert(
-      weights_size.end(),
-      options.kernel_size()->begin(),
-      options.kernel_size()->end());
-  AT_ASSERT(weights_size.size() == 2 + options.kernel_size()->size());
-
-  weight = this->register_parameter("weight", torch::empty(weights_size));
-  if (options.with_bias()) {
-    bias = this->register_parameter(
-        "bias", torch::empty(options.output_channels()));
+
+  if (options.bias()) {
+    bias = this->register_parameter("bias", torch::empty({*options.out_channels()}));
+  } else {
+    this->register_parameter("bias", Tensor(), /*requires_grad=*/false);
   }
 
-  const auto number_of_features = std::accumulate(
-      options.kernel_size()->begin(),
-      options.kernel_size()->end(),
-      options.input_channels(),
-      std::multiplies<int64_t>{});
-  const auto stdv = 1.0 / std::sqrt(number_of_features);
-  NoGradGuard no_grad;
-  for (auto& p : this->parameters()) {
-    p.uniform_(-stdv, stdv);
+  init::kaiming_uniform_(weight, /*a=*/std::sqrt(5));
+
+  if (bias.defined()) {
+    int64_t fan_in, fan_out;
+    std::tie(fan_in, fan_out) = init::_calculate_fan_in_and_fan_out(weight);
+    auto bound = 1 / std::sqrt(fan_in);
+    init::uniform_(bias, -bound, bound);
   }
 }
 
 template <size_t D, typename Derived>
 void ConvImpl<D, Derived>::pretty_print(std::ostream& stream) const {
   stream << "torch::nn::Conv" << D << "d"
-         << "(input_channels=" << options.input_channels()
-         << ", output_channels=" << options.output_channels()
-         << ", kernel_size=" << options.kernel_size()
-         << ", stride=" << options.stride() << ")";
+         << "(" << *options.in_channels()
+         << ", " << *options.out_channels()
+         << ", kernel_size=" << *options.kernel_size()
+         << ", stride=" << options.stride();
+  if (*options.padding() != *ExpandingArray<D>(0)) {
+    stream << ", padding=" << options.padding();
+  }
+  if (*options.dilation() != *ExpandingArray<D>(1)) {
+    stream << ", dilation=" << options.dilation();
+  }
+  if (*options.output_padding() != *ExpandingArray<D>(0)) {
+    stream << ", output_padding=" << options.output_padding();
+  }
+  if (options.groups() != 1) {
+    stream << ", groups=" << options.groups();
+  }
+  if (!options.bias()) {
+    stream << ", bias=" << std::boolalpha << false;
+  }
+  if (!c10::get_if<enumtype::kZeros>(&options.padding_mode())) {
+    stream << ", padding_mode=" << enumtype::get_enum_name(options.padding_mode());
+  }
+  stream << ")";
 }
 
+Conv1dImpl::Conv1dImpl(
+    ConvOptions<1> options_)
+    : ConvImpl(std::move(options_).transposed(false).output_padding(0)) {}
```

Comment:
It is because we pass `options_` by value here, as a result we can use `std::move` as an optimization. The reason for passing `options_` by value is that we want to be able to change `options_` in-place (updating its `transposed_` and `output_padding_` fields) before passing it to the `ConvImpl` constructor.

Reply:
Correct me if I'm wrong, but std::move() casts options_ to rvalue-reference, but I don't see any move-ctor or move assignment operators that can accept rvalue-references, neither in ConvOptions nor in ConvImpl

Reply:
The move-ctor for ConvOptions will be created by the compiler automatically. Per https://en.cppreference.com/w/cpp/language/move_constructor:
> If no user-defined move constructors are provided for a class type (struct, class, or union), and all of the following is true:
> ~ there are no user-declared copy constructors;
> ~ there are no user-declared copy assignment operators;
> ~ there are no user-declared move assignment operators;
> ~ there are no user-declared destructors;
> then the compiler will declare a move constructor as a non-explicit inline public member of its class with the signature T::T(T&&).

Reply:
ConvImpl has a constructor that accepts `const ConvOptions<D>&`, and we can pass an rvalue as a const reference (https://stackoverflow.com/questions/36102728/why-is-it-allowed-to-pass-r-values-by-const-reference-but-not-by-normal-referenc) :D